cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java

```diff
@@ -350,25 +385,98 @@ public void apply() throws Exception {
   /**
    * Method to add version in StreamSizeSchedule row key in SchedulerStore.
    *
-   * @throws Exception
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws DatasetManagementException
    */
-  public void upgrade()
-    throws InterruptedException, TransactionFailureException, IOException, DatasetManagementException {
-    initialize();
-    factory.createExecutor(ImmutableList.of((TransactionAware) table))
-      .execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() {
-          upgradeVersionKeys();
+  public void upgrade() throws InterruptedException, IOException, DatasetManagementException {
+    // Wait until the store is initialized
+    while (!storeInitialized.get()) {
+      TimeUnit.SECONDS.sleep(10);
+    }
+
+    final AtomicBoolean upgradeComplete = new AtomicBoolean(false);
+    try {
+      factory.createExecutor(ImmutableList.of((TransactionAware) table))
+        .execute(new TransactionExecutor.Subroutine() {
+          @Override
+          public void apply() throws Exception {
+            byte[] result = table.get(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'));
+            if (result != null) {
```

Comment:
Why should it be compared with current version? What will happen in the next version of CDAP?